app/src/main/java/org/schabi/newpipe/player/datasource/YoutubeDataSourceResolver.java

```diff
@@ -0,0 +1,234 @@
+     *                         {@link DataSpec} which will be returned by
+     *                         {@link #resolveDataSpec(DataSpec)}
+     */
+    private void addYoutubePlaybackHeaders(@NonNull final DataSpec originalDataSpec,
+                                           @NonNull final Map<String, String> httpHeaders) {
+        final String url = originalDataSpec.uri.toString();
+        if (isAndroidStreamingUrl(url)) {
+            httpHeaders.put(HttpHeaders.USER_AGENT, getAndroidUserAgent(null));
+        } else if (isIosStreamingUrl(url)) {
+            httpHeaders.put(HttpHeaders.USER_AGENT, getIosUserAgent(null));
+        } else if (isWebStreamingUrl(url) || isTvHtml5SimplyEmbeddedPlayerStreamingUrl(url)) {
+            httpHeaders.put(HttpHeaders.ORIGIN, YOUTUBE_BASE_URL);
+            httpHeaders.put(HttpHeaders.REFERER, YOUTUBE_BASE_URL);
+            httpHeaders.put(HttpHeaders.SEC_FETCH_DEST, "empty");
+            httpHeaders.put(HttpHeaders.SEC_FETCH_MODE, "cors");
+            httpHeaders.put(HttpHeaders.SEC_FETCH_SITE, "cross-site");
+        }
+    }
+
+    /**
+     * Append the {@code rn} parameter to {@code videoplayback} streaming URLs, if enabled and
+     * needed.
+     *
+     * <p>
+     * To fetch its contents, YouTube use range requests which append a {@code range} parameter or
+     * path to {@code videoplayback} URLs instead of the {@code Range} header (even if the server
+     * respond correctly when requesting a range of a resource with the header).
+     * </p>
+     *
+     * <p>
+     * This method supports both path and query parameters.
+     * </p>
+     *
+     * @param originalDataSpec the original {@link DataSpec} provided by ExoPlayer in
+     *                         {{@link #resolveDataSpec(DataSpec)}}
+     * @param dataSpecBuilder  the {@link DataSpec.Builder} used to modify the original
+     *                         {@link DataSpec}
+     * @return the range parameter to add, or an empty string
+     */
+    @NonNull
+    private String appendRangeParameterIfNeeded(@NonNull final DataSpec originalDataSpec,
+                                                @NonNull final DataSpec.Builder dataSpecBuilder) {
+        if (rangeParameterEnabled) {
+            final String rangeParameter = buildRangeParameter(
+                    originalDataSpec.position, originalDataSpec.length);
+            if (rangeParameter != null) {
+                // Workaround: prevent ExoPlayer HttpDataSources sending a Range HTTP header by
+                // setting the DataSpec position to 0 and its length to unset
+                // See HttpUtil#buildRangeRequestHeader(long, long)
+                // If the content is served with compression, this workaround may make ExoPlayer
+                // not work properly
+                dataSpecBuilder.setPosition(0);
+                dataSpecBuilder.setLength(C.LENGTH_UNSET);
```

Comment:
What happens if you were to remove these two lines? You would start sending the range header, but why is that unwanted? According to what you wrote above: "the server responds correctly when requesting a range of a resource with the header".

Reply:
We would send the HTTP header and the query parameter in this case, which is something uncommon and could be used by YouTube to detect us.

YouTube clients send either the header (rarely used) or the query parameter (mostly used).